### PR TITLE
[dev] moe(fix): Avoid TE cuda graph dummy attention masks

### PR DESCRIPTION
**Pull request by yuzhongw-nvidia**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

PR for dev: https://github.com/NVIDIA/Megatron-LM/pull/5173

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Summary

  This PR avoids allocating dummy attention masks for TransformerEngine CUDA graph attention when attention masks are intentionally omitted by the dataloader.

  When `--no-create-attention-mask-in-dataloader` is used, the eager forward path passes `attention_mask=None`. The TE CUDA graph path, however, was still creating synthetic all-zero masks for capture/replay with shape `[micro_batch, 1, local_seq_len, global_seq_len]`. With context parallelism, scaling `seq_len` and CP together keeps local sequence length constant, but the dummy mask still scaled with global `seq_len`, causing unnecessary CUDA graph memory growth.

  ## Changes

  - Plumb `create_attention_mask_in_dataloader` into `TransformerConfig`.
  - For TE CUDA graph attention with `create_attention_mask_in_dataloader=False`, omit `attention_mask` from both static capture inputs and replay kwargs instead of allocating a dummy mask.
  - Preserve the previous mask behavior when dataloader-created masks are enabled.
  - Add warnings for attention mask types where omitting the mask may be unsafe.
  - Update unit coverage for TE CUDA graph replay/static input behavior.

## Validation

  Validated on a Blackwell platform using a 4-layer Qwen3-30B debug recipe with partial CUDA graph enabled.

  The key comparison is the same sequence-length / CP cases before vs. after the fix:

  | Case | Before fix peak allocated | Before fix peak reserved | Before fix device memory | After fix peak allocated | After fix peak reserved | After fix device memory |
  | --- | ---: | ---: | ---: | ---: | ---: | ---: |
  | 16k-CP1 | 29,258 MB | 32,702 MB | 41,219 MB | 29,258 MB | 31,942 MB | 40,459 MB |
  | 32k-CP2 | 29,263 MB | 34,666 MB | 44,265 MB | 29,263 MB | 32,424 MB | 42,023 MB |
  | 64k-CP4 | 31,128 MB | 36,612 MB | 46,601 MB | 29,264 MB | 32,626 MB | 42,621 MB |

  For the long-sequence cases, the fix removes the CUDA-graph-specific dummy-mask memory:

  | Case | Peak allocated reduction | Peak reserved reduction | Device memory reduction |
  | --- | ---: | ---: | ---: |
  | 16k-CP1 | 0 MB | 760 MB | 760 MB |
  | 32k-CP2 | 0 MB | 2,242 MB | 2,242 MB |
  | 64k-CP4 | 1,864 MB | 3,986 MB | 3,980 MB |

  The growth from `16k-CP1 -> 64k-CP4` is also reduced:

  - Before fix: device memory increased by `5,382 MB`.
  - After fix: device memory increased by `2,162 MB`.

  The remaining delta matches the no-CG / MoE-only CP baseline. This confirms that the attention CUDA graph path was retaining synthetic `[local_seq_len, global_seq_len]` attention masks, and that omitting those masks when `--no-create-attention-mask-in-dataloader` is set removes the long-sequence CP memory regression.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by yaox12 — APPROVED** [review]

File: megatron/core/transformer/transformer_layer.py [hunk 1]
```diff
@@ -1063,6 +1069,21 @@ def get_layer_static_inputs(self, seq_length, micro_batch_size):
 
         return static_inputs
 
+    def _should_omit_attention_mask_for_te_cuda_graph(self):
```
Comment (buptzyb) [inline megatron/core/transformer/transformer_layer.py]: why not just return not self.config.create_attention_mask_in_dataloader?

Comment (yuzhongw-nvidia) [reply megatron/core/transformer/transformer_layer.py]: Thanks. Resolved.

**Comment by yuzhongw-nvidia** [comment]
/ok to test [46faa66](https://github.com/NVIDIA/Megatron-LM/commit/46faa6637c2d6726c0e59b0a0f5db37689d29e04)

**Review by buptzyb — APPROVED** [review]

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/c046dcce78f13e7255791a90ac82f628c967c35b

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/3671459e3cbfd773ddd40b12a2a1a6670ca33778
